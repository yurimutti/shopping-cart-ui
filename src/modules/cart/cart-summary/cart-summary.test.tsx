import { render, screen, fireEvent } from "@testing-library/react";
import { describe, it, expect, vi } from "vitest";
import { CartSummary } from "./";
import type { Product } from "@/types/product";

const mockProduct: Product = {
  id: 1,
  name: "Cheeseburger Combo",
  price: 12.99,
  quantity: 2,
};

describe("<CartSummary />", () => {
  const defaultProps = {
    cart: [mockProduct],
    onRemove: vi.fn(),
    onAdd: vi.fn(),
    onCheckout: vi.fn(),
  };

  it("should render cart items correctly", () => {
    const { container } = render(<CartSummary {...defaultProps} />);

    expect(screen.getByText(mockProduct.name)).toBeInTheDocument();
    expect(
      screen.getByText(`Quantity: ${mockProduct.quantity}`)
    ).toBeInTheDocument();
    expect(screen.getByText(`$${mockProduct.price.toFixed(2)}`)).toBeInTheDocument();
    expect(screen.getByText(`$${(mockProduct.price * mockProduct.quantity).toFixed(2)}`)).toBeInTheDocument();

    expect(
      screen.getByRole("button", { name: "Checkout" })
    ).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render empty state when cart is empty", () => {
    render(
      <CartSummary {...defaultProps} cart={[]} />
    );

    expect(screen.getByText("Your cart is empty")).toBeInTheDocument();
    expect(screen.getByRole("button", { name: "Checkout" })).toBeDisabled();
  });

  it("should call onAdd when add button is clicked", () => {
    render(<CartSummary {...defaultProps} />);

    const addButton = screen.getByRole("button", {
      name: `Add one ${mockProduct.name}`,
    });

    fireEvent.click(addButton);
    expect(defaultProps.onAdd).toHaveBeenCalledWith(mockProduct);
  });

  it("should call onRemove when remove button is clicked", () => {
    render(<CartSummary {...defaultProps} />);

    const removeButton = screen.getByRole("button", {
      name: `Remove ${mockProduct.name}`,
    });

    fireEvent.click(removeButton);
    expect(defaultProps.onRemove).toHaveBeenCalledWith(mockProduct.id);
  });

  it("should call onRemove when decrement button is clicked", () => {
    render(<CartSummary {...defaultProps} />);

    const decrementButton = screen.getByRole("button", {
      name: `Remove one ${mockProduct.name}`,
    });

    fireEvent.click(decrementButton);
    expect(defaultProps.onRemove).toHaveBeenCalledWith(mockProduct.id);
  });

  it("should call onCheckout when checkout button is clicked", () => {
    render(<CartSummary {...defaultProps} />);

    const checkoutButton = screen.getByRole("button", { name: "Checkout" });
    fireEvent.click(checkoutButton);

    expect(defaultProps.onCheckout).toHaveBeenCalled();
  });
});
