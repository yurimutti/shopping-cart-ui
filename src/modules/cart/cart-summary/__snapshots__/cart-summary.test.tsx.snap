// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<CartSummary /> > should render cart items correctly 1`] = `
<div
  class="flex flex-col gap-4"
>
  <ul
    aria-label="Shopping cart items"
    class="space-y-2 mb-4 min-h-[100px] max-h-[600px] overflow-y-auto"
    role="list"
  >
    <li
      class="flex justify-between items-center bg-white p-4 rounded-lg border border-gray-200 shadow-sm"
    >
      <div
        class="flex flex-col gap-1"
      >
        <span
          class="font-medium text-black"
        >
          Cheeseburger Combo
        </span>
        <span
          class="text-gray-500 text-sm"
        >
          Quantity: 
          2
        </span>
        <span
          class="text-gray-700 text-sm"
        >
          $
          12.99
        </span>
      </div>
      <div
        class="flex items-center gap-2"
      >
        <button
          aria-label="Remove Cheeseburger Combo"
          class="text-gray-500 hover:text-black text-sm font-medium"
        >
          Remove
        </button>
        <button
          aria-label="Add one Cheeseburger Combo"
          class="bg-black hover:bg-gray-800 text-white rounded-full w-7 h-7 flex items-center justify-center"
        >
          +
        </button>
        <button
          aria-label="Remove one Cheeseburger Combo"
          class="bg-black hover:bg-gray-800 text-white rounded-full w-7 h-7 flex items-center justify-center"
        >
          -
        </button>
      </div>
    </li>
  </ul>
  <div
    class="flex justify-between items-center mb-4 text-base font-medium text-black"
  >
    <span>
      Total:
    </span>
    <span>
      $
      25.98
    </span>
  </div>
  <button
    class="w-full rounded-full p-3 font-semibold transition-colors bg-red-600 hover:bg-red-700 text-white"
  >
    Checkout
  </button>
</div>
`;
